### PR TITLE
Use the maven-publish plugin to publish AAR files.

### LINKED ISSUE #101083: Build AARs with Gradle 7.0
Gradle 7.0 removed `org.gradle.api.artifacts.maven.MavenDeployer` currently used to build AARs.

There's a new Maven Publish Plugin that can be used to output the artifacts to a local destination:
https://docs.gradle.org/current/userguide/publishing_maven.html#publishing_maven:install

### PR DESCRIPTION
**Pull request by AesSedai101**
This PR changes `flutter build aar` to make use of the [maven-publish](https://docs.gradle.org/current/userguide/publishing_maven.html) plugin, removing the `maven` plugin which as been previously deprecated and removed in Gradle 7.

Fixes #101083 

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by AesSedai101** [comment]
Builds are currently failing because `maven-publish` creates different Android variants as _classifiers_ (`io.flutter.devicelab.hello:flutter:1.0:debug`) while the old `publish` plugin created them as different artefacts with the variant appended to the `artifactId` (`io.flutter.devicelab.hello:flutter_debug:1.0`).

**Comment by blasten** [comment]
you may also need to change:
https://github.com/flutter/flutter/blob/0b97874895eb3e14c1cf0e65bd7fca3a17c62b02/packages/flutter_tools/lib/src/android/gradle_utils.dart#L32-L33

Maybe, remove `templateAndroidGradlePluginVersionForModule` and replace with `templateDefaultGradleVersion`.

File: AUTHORS [hunk 1]
```diff
@@ -93,3 +93,4 @@ Alberto Miola <betoman96@gmail.com>
 Twin Sun, LLC <google-contrib@twinsunsolutions.com>
 Taskulu LDA <contributions@taskulu.com>
 Jonathan Joelson <jon@joelson.co>
+Elsabe Ros <hello@elsabe.dev>
```
Comment (blasten) [inline AUTHORS:96]: please add an end of file

File: packages/flutter_tools/gradle/aar_init_script.gradle [hunk 2]
```diff
@@ -34,13 +30,37 @@ void configureProject(Project project, String outputDir) {
...
+                url = uri("file://${outputDir}/outputs/repo")
+            }
+        }
+    }
+
+
+    /*
+    Some extra work to keep the publications in the same format as the previous plugin. Instead of
+    using classifiers for the variants, append `_{variant}` to the artifactId
+     */
```
Comment (blasten) [inline packages/flutter_tools/gradle/aar_init_script.gradle:45]: for consistency, use `//`

File: packages/flutter_tools/gradle/aar_init_script.gradle [hunk 3]
```diff
@@ -135,6 +145,15 @@ Set<Configuration> flattenConfiguration(Configuration configuration) {
     return configs
 }
 
+/*
+maven-publish has to be applied _before_ the project gets evaluated, but some of the code in
+`configureProject` requires the project to be evaluated. Apply the maven plugin to all projects, but
+only configure it if it matches the conditions in `projectsEvaluated`
+*/
```
Comment (blasten) [inline packages/flutter_tools/gradle/aar_init_script.gradle:152]: for consistency, use `//`

**Comment by blasten** [comment]
Based on https://github.com/gradle/gradle/issues/12324#issuecomment-868474341, I think the issue appears to be that we can't configure multiple publications at once. I wonder if it could just configure the task that was currently requested from Dart, and this way, just configure a single publication.

**Comment by AesSedai101** [comment]
> Based on [gradle/gradle#12324 (comment)](https://github.com/gradle/gradle/issues/12324#issuecomment-868474341), I think the issue appears to be that we can't configure multiple publications at once. I wonder if it could just configure the task that was currently requested from Dart, and this way, just configure a single publication.

So that is kind of what I'm trying to do with `project.components.forEach { component ->`, which does the trick - it produces `_debug`, `_profile` and `_release` artifactId's. The next step is to convince gradle to resolve those correctly

**Comment by AesSedai101** [comment]
Ok, I'm not finding a way to get the maven-publish plugin to resolve these dependencies correctly. What I would like to propose is to put a [relocation pom](https://docs.gradle.org/current/userguide/publishing_maven.html#publishing_maven:relocation) at the old coordinates and then publish the actual artefacts as per the new gradle way (one artefact with classifiers). This way, we keep backward compatibility, but still publish in the "default" way

**Comment by blasten** [comment]
@AesSedai101 makes sense :) Thank you for doing all the research!

**Comment by AesSedai101** [comment]
> @AesSedai101 makes sense :) Thank you for doing all the research!

perfect! I'll make the changes and update the tests sometimes today 🙂

File: packages/flutter_tools/lib/src/project.dart [hunk 4]
```diff
@@ -589,7 +589,7 @@ class AndroidProject extends FlutterProjectPlatform {
         'agpVersion': gradle.templateAndroidGradlePluginVersion,
         'kotlinVersion': gradle.templateKotlinGradlePluginVersion,
         'gradleVersion': gradle.templateDefaultGradleVersion,
-        'gradleVersionForModule': gradle.templateDefaultGradleVersionForModule,
+        'gradleVersionForModule': gradle.templateAndroidGradlePluginVersion,
```
Comment (AesSedai101) [inline packages/flutter_tools/lib/src/project.dart:592]:
```suggestion
        'gradleVersionForModule': gradle.templateDefaultGradleVersion,
```

**Comment by AesSedai101** [comment]
The `build_aar_module_test`s appears to be failing because of jcenter connection issues

File: .ci.yaml [hunk 5]
```diff
@@ -188,13 +188,13 @@ targets:
       caches: >-
         [
           {"name":"gradle","path":"gradle"},
-          {"name": "openjdk", "path": "java"}
+          {"name": "openjdk", "path": "java11"}
```
Comment (godofredoc) [inline .ci.yaml:191]: Name is the one that needs to change, path should be kept as java.

Comment (blasten) [reply .ci.yaml:191]: there are many caches in this file that have this entry: `{"name": "openjdk", "path": "java11"}`. Are those set incorrectly?

Comment (godofredoc) [reply .ci.yaml:191]: Yes the are all incorrect, the modules will add cache/java to the path and this cache won't have any effect.

Comment (blasten) [reply .ci.yaml:191]: got it. can the name be any arbitrary string?

Comment (godofredoc) [reply .ci.yaml:191]: we are usually using dependency_<version> for named caches.

Comment (blasten) [reply .ci.yaml]: Done

**Review by blasten — APPROVED** [review]
Nicely done, thank you! LGTM

**Review by godofredoc — APPROVED** [review]

**Comment by JunhuaLin** [comment]
https://github.com/flutter/flutter/issues/111643

@AesSedai101

**Pull request merged**